GitUI/UserControls/RevisionGrid/Graph/RevisionGraph.cs

```diff
@@ -290,13 +262,67 @@ public void Add(GitRevision revision, int? insertScore = null, int insertRange =
 
             // Ensure all parents are loaded before adding it to the _nodes list. This is important for ordering.
             ImmutableInterlocked.Update(ref _nodes, (list, revision) => list.Add(revision), revisionGraphRevision);
+            MarkCacheAsInvalidIfNeeded(minExistingScore);
+        }
 
-            if (!updateParents)
+        /// <summary>
+        /// Insert workTree and index revisions to the graph, offset existing revisions.
+        /// </summary>
+        /// <param name="workTreeRev">The workTree revision to add.</param>
+        /// <param name="indexRev">The index revision to add.</param>
+        /// <param name="parents">Parent ids for the revision to find (and insert before).</param>
+        public void Insert(GitRevision workTreeRev, GitRevision indexRev, IReadOnlyList<ObjectId> parents)
+        {
+            // Find the first parent that is already in the graph, insert a gap in scores before.
+            // Set the default insert score if not found as before the existing.
+            int insertScore = int.MinValue;
+            foreach (ObjectId parentId in parents!)
             {
-                // The rows may already be cached, invalidate and update but do not reload rows in this context
-                _reorder = true;
-                Updated?.Invoke();
+                if (TryGetNode(parentId, out RevisionGraphRevision? parentRev))
+                {
+                    // Insert before this ancestor
+                    const int insertRange = 2;
+                    int limitScore = parentRev.Score;
+                    insertScore = limitScore - insertRange;
+                    foreach (RevisionGraphRevision graphRevision in _nodeByObjectId.Values)
+                    {
+                        if (graphRevision.Score < limitScore)
+                        {
+                            // Lower existing scores to reserve the inserted range
+                            graphRevision.OverrideScore(graphRevision.Score - insertRange);
+                        }
+                    }
+
+                    break;
+                }
             }
+
+            // Add graph revisions including segment.
+            RevisionGraphRevision workTreeGraphRevision = new(workTreeRev.ObjectId, insertScore)
+            {
+                GitRevision = workTreeRev
+            };
+
+            RevisionGraphRevision indexGraphRevision = new(indexRev.ObjectId, ++insertScore)
+            {
+                GitRevision = indexRev
+            };
+
+            workTreeGraphRevision.AddParent(indexGraphRevision, workTreeGraphRevision.Score + 1);
+            _nodeByObjectId.TryAdd(workTreeRev.ObjectId, workTreeGraphRevision);
+            _nodeByObjectId.TryAdd(indexRev.ObjectId, indexGraphRevision);
```

Comment:
```suggestion
            _nodeByObjectId.Add(workTreeRev.ObjectId, workTreeGraphRevision);
            _nodeByObjectId.Add(indexRev.ObjectId, indexGraphRevision);
```
If this can really be attempted twice, then skip the entire block "// Add graph revisions including segment." using
`if (!_nodeByObjectId.TryAdd(workTreeRev.ObjectId))`
or 
```
            if (_nodeByObjectId.TryAdd(workTreeRev.ObjectId, workTreeGraphRevision))
            {
                RevisionGraphRevision indexGraphRevision = new(indexRev.ObjectId, ++insertScore)
                .
                .
                .
                ImmutableInterlocked.Update(ref _nodes, (list, revision) => list.Add(revision), indexGraphRevision);
            }
```

Reply:
If added twice, the revision will be replaced but a segment added earlier may fail - but that would be very strange.
No checks here, to not delay loading

Update of _nodes should be done last, that is used to add to nodescache to update grid.
In #11268 _nodes is removed so the code is slightly changed

Do you want to remove the comment?

Reply:
The PR implementation creates an inconsistent state:
- old RevisionGraphRevision is kept in `_nodeByObjectId`
- new RevisionGraphRevision in `_nodes`

The following avoids the inconsistency and does not add extra cost (except an `if` on the return value):
```
            // Add graph revisions including segment.
            RevisionGraphRevision workTreeGraphRevision = new(workTreeRev.ObjectId, insertScore)
            {
                GitRevision = workTreeRev
            };

            if (_nodeByObjectId.TryAdd(workTreeRev.ObjectId, workTreeGraphRevision)
            {
                RevisionGraphRevision indexGraphRevision = new(indexRev.ObjectId, ++insertScore)
                {
                    GitRevision = indexRev
                };

                workTreeGraphRevision.AddParent(indexGraphRevision, workTreeGraphRevision.Score + 1);
                _nodeByObjectId.Add(indexRev.ObjectId, indexGraphRevision);
                ImmutableInterlocked.Update(ref _nodes, (list, revision) => list.Add(revision), workTreeGraphRevision);
                ImmutableInterlocked.Update(ref _nodes, (list, revision) => list.Add(revision), indexGraphRevision);
            }
```

Reply:
But worktree and index is not added previously - the code should return if that is checked.
Add segment should b done before.

_nodeByObjectId.Add() will be the latest lines in next PR anyway.

Reply:
> _nodeByObjectId.Add() will be the latest lines in next PR anyway.

But actually, it still is a blind `_nodeByObjectId.TryAdd()` ignoring the result, which is a code smell.
Either "But worktree and index is not added previously" is true or we should not call the entire method, should we?
At least it should be
```suggestion
            _ = _nodeByObjectId.TryAdd(workTreeRev.ObjectId, workTreeGraphRevision)
                && _nodeByObjectId.TryAdd(indexRev.ObjectId, indexGraphRevision);
```

Reply:
Adjusted to the suggestion

There is no Add() for CuncurrentDirectory.
There is AddOrUpdate() with an update method though.

_nodeByObjectId is not checked for normal insertion either (before or now)
A Debug.Assert(!Contains()) could be used here

But I rather save this suggestion to next PR where structures are changed.